common/base/tests/Test_UtilsThreadsafeContainers.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include "gtest/gtest.h"
+
+#include <claragenomics/utils/threadsafe_containers.hpp>
+
+#include <algorithm>
+#include <mutex>
+#include <cstdint>
+#include <numeric>
+#include <thread>
+
+namespace claragenomics
+{
+
+// *** test ThreadsafeDataProvider ***
+
+void test_threadsafe_data_provider(const std::size_t number_of_elements,
+                                   const std::size_t number_of_threads)
```

Comment:
Optional: I would use `int32_t` instead of `std::size_t` for everything in this code. None of these variables refer to memory sizes or huge position values.

Reply:
I agree. The reason I was using `std::size` is that some of those values are used to index a vector which takes `std::size_t`, but I guess it really does not make a big difference here

Reply:
done

Reply:
Great, thanks!
To resolve the whole "prefer signed integers for numbers" vs. "`std::vector::operator[]` takes `size_t`" problem, Cpp Core Guidelines actually recommends to use `gsl::index`[1]. It allows to have a signed index which is compatible to `size_t`.
We may discuss switching to that, but I it seems a bit silly to have another dependency just for this purpose. So I would just stick with using signed integers everywhere (+ `get_size()`) and rely on implicit conversion to `size_t`.
[1] https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-subscripts